Hummingbird.docc/Articles/Recommendations.md

```diff
@@ -0,0 +1,342 @@
+
+Let's take **UserController.swift** as an example:
+
+```swift
+struct UserController<Context: RequestContext> {
+    var routes: RouteCollection<Context> {
+        return RouteCollection(context: Context.self)
+            .get { request, context in
+                ...
+            }
+    }
+}
+```
+
+Controllers can specify dependencies on services they need, like a database connection. To do so, add these properties on the controller:
+
+```swift
+struct UserController<Context: RequestContext> {
+    // Dependency on MongoDB
+    let database: MongoDatabase
+
+    var routes: RouteCollection<Context> {
+        return RouteCollection(context: Context.self)
+            .get { request, context in
+                ...
+            }
+    }
+}
+```
+
+If you need the ability to inject dependencies, you can use a `protocol` to abstract the implementation of the dependency.
+
+```swift
+protocol SomeInjectableService {
+    func doSomething()
+}
+
+struct UserController<Context: RequestContext> {
+    let service: any SomeInjectableService
+
+    var routes: RouteCollection<Context> {
+        return RouteCollection(context: Context.self)
+            .get { request, context in
+                self.service.doSomething()
+                ...
+            }
+    }
+}
+```
+
+All Controllers' routes are combined into a single router in **BuildRouter.swift**:
+
+```swift
+let router = Router(context: MyContext.self)
+router.add("users", routes: UserController().routes)
+```
+
+## Request Context
+
+It's also recommended to users that they refine the AppContext requirements for only as much as strictly required by the routes. For example:
+
+```swift
+// ✅ This is recommended, because it allows a different RequestContext to be used for login routes
+struct UserController<Context: RequestContext> { ... }
+```
+
+If a route has requirements for the request, like being authenticated, it's recommended to leverage _protocol composition_ to refine the RequestContext.
+
+```swift
+// ✅ This is recommended, because it leverages protocol composition to refine the RequestContext
+// while still being flexible enough to allow different concrete types to fulfil the requirements
+struct UserProfileController<
+    Context: RequestContext & AuthenticatedRequestContext
+> { ... }
+```
+
+Users should define various traits related to a request's state (context) in separate protocols.
+
+```swift
+protocol RateLimitedRequestContext {
+    var rateLimitStatus: RateLimitStatus { get }
+    var ipAddress: String { get }
+}
+
+protocol AuthenticatedRequestContext {
+    var user: User { get }
+}
+```
+
+## Middleware
+
+Middleware follow the same RequestContext conventions and recommendations as Routes. There's one distinction between Routes in that middleware can modify the request context.
```

Comment:
This should be made much clearer. In many cases Middleware are the re-usable elements of a project. Using generic parameters and defining requirements using protocols is much more important here.